src/include/cart/api.h

```diff
@@ -685,98 +685,154 @@ crt_ep_abort(crt_endpoint_t *ep);
+#define CRT_GEN_GET_TYPE(seq) BOOST_PP_SEQ_ELEM(0, seq)
 #define CRT_GEN_GET_NAME(seq) BOOST_PP_SEQ_ELEM(1, seq)
-#define CRT_GEN_GET_KIND(seq) BOOST_PP_SEQ_TAIL(BOOST_PP_SEQ_TAIL(seq))
+#define CRT_GEN_GET_KIND(seq) BOOST_PP_SEQ_ELEM(2, seq)
 
-#define CRT_GEN_STRUCT_FIELD(r, data, seq)				\
-	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_ARRAY, CRT_GEN_GET_KIND(seq)),	\
+/* convert constructed name into proper name */
+#define crt_proc_struct BOOST_PP_RPAREN() BOOST_PP_CAT BOOST_PP_LPAREN() \
+	crt_proc_struct_,
+
+#define CRT_GEN_X(x) x
+#define CRT_GEN_X2(x) CRT_GEN_X BOOST_PP_LPAREN() crt_proc_##x BOOST_PP_RPAREN()
+#define CRT_GEN_GET_FUNC(seq) CRT_GEN_X2 BOOST_PP_SEQ_FIRST_N(1, seq)
+#define CRT_GEN_X3(x) CRT_GEN_X BOOST_PP_LPAREN() fail_##x BOOST_PP_RPAREN()
+#define CRT_GEN_FAIL_LABEL(seq) CRT_GEN_X3 BOOST_PP_SEQ_TAIL(BOOST_PP_SEQ_FIRST_N(2, seq))
+
+#define CRT_GEN_STRUCT_FIELD(seq)					\
+	BOOST_PP_EXPAND(						\
+	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_GEN_X CRT_ARRAY, CRT_GEN_GET_KIND(seq)),\
 		struct {						\
 			uint64_t		 ca_count;		\
 			CRT_GEN_GET_TYPE(seq)	*ca_arrays;		\
 		},							\
 		CRT_GEN_GET_TYPE(seq))					\
-	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_PTR, CRT_GEN_GET_KIND(seq)),	\
-		*CRT_GEN_GET_NAME(seq), CRT_GEN_GET_NAME(seq));
+	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_GEN_X CRT_PTR, CRT_GEN_GET_KIND(seq)), \
+		*CRT_GEN_GET_NAME(seq), CRT_GEN_GET_NAME(seq));)
+
+#define CRT_GEN_STRUCT_FIELDS(z, n, seq)				\
+	CRT_GEN_STRUCT_FIELD(BOOST_PP_SEQ_ELEM(n, seq))
 
 #define CRT_GEN_STRUCT(struct_type_name, seq)				\
 	struct struct_type_name {					\
-		BOOST_PP_SEQ_FOR_EACH(CRT_GEN_STRUCT_FIELD, , seq)	\
+		BOOST_PP_REPEAT(BOOST_PP_SEQ_SIZE(seq),			\
+				CRT_GEN_STRUCT_FIELDS, seq)		\
 	};
 
-/* convert constructed name into proper name */
-#define crt_proc_struct BOOST_PP_RPAREN() BOOST_PP_CAT BOOST_PP_LPAREN() \
-	crt_proc_struct_,
-
-#define CRT_GEN_X(x) x
-#define CRT_GEN_X2(x) CRT_GEN_X BOOST_PP_LPAREN() crt_proc_##x BOOST_PP_RPAREN()
-#define CRT_GEN_GET_FUNC(seq) CRT_GEN_X2 BOOST_PP_SEQ_FIRST_N(1, seq)
-
-#define CRT_GEN_PROC_FIELD(r, ptr, seq)					\
-	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_ARRAY, CRT_GEN_GET_KIND(seq)),	\
-	{								\
-		uint64_t count = ptr->CRT_GEN_GET_NAME(seq).ca_count;	\
-		CRT_GEN_GET_TYPE(seq)** e_ptrp = &ptr->CRT_GEN_GET_NAME(seq).ca_arrays; \
-		CRT_GEN_GET_TYPE(seq)* e_ptr = ptr->CRT_GEN_GET_NAME(seq).ca_arrays; \
-		int i;							\
-		/* process the count of array first */			\
-		rc = crt_proc_uint64_t(proc, proc_op, &count);		\
-		if (unlikely(rc))					\
-			goto out;					\
-		ptr->CRT_GEN_GET_NAME(seq).ca_count = count;		\
-		if (unlikely(count == 0)) {				\
-			if (DECODING(proc_op))				\
-				*e_ptrp = NULL;				\
-			goto next_field_##r;				\
-		}							\
-		if (DECODING(proc_op)) {				\
-			D_ALLOC_ARRAY(e_ptr, (int)count);		\
-			if (unlikely(e_ptr == NULL)) {			\
-				rc = -DER_NOMEM;			\
-				goto out;				\
-			}						\
-			*e_ptrp = e_ptr;				\
+#define CRT_GEN_PROC_ARRAY(ptr, seq)					\
```

Comment:
(style)  Macros with flow control statements should be avoided
(style)  do {} while (0) macros should not be semicolon terminated